### DISCUSSION TITLE
Would there be interest in a multi-JavaScript engine Jasmine runner for ECMAScript?

### DISCUSSION THREAD

**Discussion by ajvincent** [jasmine/jasmine#2047] [General]
### Your question

I'm considering creating a custom Jasmine runner project leveraging [eshost](https://github.com/tc39/eshost) to run pure-JavaScript tests against multiple JavaScript engines.  I will very likely be doing most of this myself, including maintenance.  I'm really just wondering if there's any interest among the Jasmine projects for something like this.

I did look around npm to see if there is an existing package for this, but didn't find any.

If the answer's no, that's fine.

### Example code that demonstrates the problem

```JavaScript
N/A
```

### jasmine-core version

5.3+

### Node.js or browser version

Node 18+

### Operating System

Windows, MacOS, Linux

**Comment by sgravrock** [comment]
I'm not sure if this is a general interest check or a question for me specifically. So I'll give my answer then leave it open for others to see and respond to.

I'm not personally interested, but I can see the potential value of such a thing. There have been sporadic efforts to run Jasmine on Rhino, Nashorn, and SpiderMonkey in the past but they mostly seemed to fizzle out from lack of sustained interest. Maybe a single tool that targets a whole bunch of engines would prove more sustainable.

There are some portability challenges, as you might already know. Even if all the engines of interest implement enough of ECMAScript to get the job done, a testing framework needs to do some things that can't (as far as I know) be done portably:

* Unhandled exception and unhandled promise rejection handling
* Stack-clearing (although it's entirely possible that just calling `queueMicrotask` is good enough in all non-browser engines)
* Understanding stack traces
* Module loader hooks (Jasmine can punt on this because the user can control the Node command line)

If you're looking to use jasmine-core in this, there's also the problem that it doesn't support most of the engines supported by eshost. Historically we've been at least somewhat willing to merge PRs that fix things for unsupported engines, and [that hasn't changed](https://github.com/jasmine/jasmine/pull/2034#pullrequestreview-2200793313). However, there's no guarantee that jasmine-core will keep working on unsupported engines.

**Comment by ajvincent** [comment]
> I'm not sure if this is a general interest check or a question for me specifically. So I'll give my answer then leave it open for others to see and respond to.

Very much general interest.  I'm just starting to plan for this.  There are two bits of good news for others:

1. I genuinely believe I will need this for my [es-membrane](https://github.com/ajvincent/es-membrane/tree/master/packages/v2-es-membrane) project.
2. I am willing to put in lots of spare time over months to build out supporting infrastructure for said project.

So it won't happen quickly with just me driving, but I do expect to put in a good faith effort.  One thing is I'm not dumb enough to try designing it in this thread.

As for alternatives:

- jest-core specifically says it's not supporting a public API yet
- test-262 might work but is designed to test engines, not libraries like mine.

> There have been sporadic efforts to run Jasmine on Rhino, Nashorn, and SpiderMonkey in the past but they mostly seemed to fizzle out from lack of sustained interest. 

Can you point me to some references I can learn from, please?

> ... a testing framework needs to do some things that can't (as far as I know) be done portably:

This is invaluable information I didn't have before.  No, I didn't already know about any of those portability challenges.  `queueMicrotask`, in particular, is _not_ part of ECMAScript, so I'll have to look for (or write) a shim.

> If you're looking to use jasmine-core in this, there's also the problem that it doesn't support most of the engines supported by eshost.

I'll have to keep that in mind.  Yes, I was planning on wrapping this around eshost and jasmine-core, with jasmine-npm as a model to follow.  Probably the best smoke-test I can imagine for this is passing the jasmine-core tests with eshost.

I really thank you for giving me these pointers, even if you are passing on this conceptually.

**Comment by sgravrock** [comment]
> > There have been sporadic efforts to run Jasmine on Rhino, Nashorn, and SpiderMonkey in the past but they mostly seemed to fizzle out from lack of sustained interest.
> 
> Can you point me to some references I can learn from, please?

There is [SpiderMonkey tooling that's still maintained](https://github.com/ptomato/jasmine-gjs/), although it doesn't seem to have ever attracted much participation and [might be stuck on Jasmine 3.x](https://github.com/ptomato/jasmine-gjs/commit/35de3f69b3c6ba9de38d223221d63d54dfd9d760). The author had to repeatedly [patch Jasmine](https://github.com/ptomato/jasmine-gjs/commits/6efc9b018cc8205d2af70f0ef74c5bc25095352e/lib/local-modifications.patch) to fix cases where GJS complained about valid JavaScript, e.g. https://github.com/jasmine/jasmine/issues/751 .

For Rhino and Nashorn, I'm not aware of any tools like jasmine-gjs. We do see some write-ups, e.g.:

https://groups.google.com/g/jasmine-js/c/Uc8iux44x_Q
https://eshepelyuk.github.io/2014/11/26/-testing-jvm-javascript-jasmine-spock-nashorn.html

Interest seems to have mostly died off by around 2015, although I see a handful of [attempts to get Jasmine working on Nashorn](https://github.com/jasmine/jasmine/issues?q=is%3Aissue+nashorn+is%3Aclosed) as late as 2017. jasmine-core 3.99 is the last version that could've possibly run on either Rhino or Nashorn, since 4.0 and later require at least ES2017.

> > ... a testing framework needs to do some things that can't (as far as I know) be done portably:
> 
> This is invaluable information I didn't have before. No, I didn't already know about any of those portability challenges. `queueMicrotask`, in particular, is _not_ part of ECMAScript, so I'll have to look for (or write) a shim.

I suppose it's a bit of an overstatement to say that you can't do stack clearing portably. You just can't get good performance in a portable way. Early Jasmine 2.x just did a setTimeout every so often. That should work everywhere, but it turns out that in any given environment there's a faster (often _much_ faster) way to do it. Jasmine's [current approach](https://github.com/jasmine/jasmine/blob/d0fe5c4712c73d63119bcbf09f5cc327926de81a/src/core/ClearStack.js) is optimized for each environment.

**Comment by ajvincent** [comment]
Having spent some time on this over the last month, I can see that error handling is indeed going to be a big problem.  The `installGlobalErrors` code in particular is not going to work at all, and it's the very first thing `execute()` does.

On the flip side, using rollup to bundle Jasmine and the specifications together with a custom reporter does appear to be viable.

UPDATE: disabling the global errors code. :-(
